### PR TITLE
Added a timeout feature to the server. 

### PR DESCRIPTION
**Pull request by hiddensudo**
If a client does not send any  message within the specified time (as configured in the configuration file), the server will disconnect the client

### CODE REVIEW

File: include/Server/include/Server.h [hunk 1]
```diff
@@ -30,11 +29,13 @@ class SmtpServer {
     size_t max_threads;
 
     boost::asio::steady_timer timeout_timer_;
+    std::chrono::seconds timeout_seconds_;
```
Comment (shevmee) [inline include/Server/include/Server.h]: According to the code style we should name variables like m_timeout_seconds

File: include/Server/include/Server.h [hunk 2]
```diff
@@ -30,11 +29,13 @@ class SmtpServer {
...
 
-  private:
+private:
     void Accept();
-    void saveData(const std::string& line, MailMessageBuilder& mail_builder,
-                  SocketWrapper& socket_wrapper, bool& in_data);
+    void saveData(const std::string& line, MailMessageBuilder& mail_builder, SocketWrapper& socket_wrapper,
```
Comment (shevmee) [inline include/Server/include/Server.h]: Rename to SaveData and delete unused params

File: include/SocketWrapper/src/SocketWrapper.cpp [hunk 3]
```diff
@@ -106,4 +106,56 @@ std::future<void> SocketWrapper::startTlsAsync(boost::asio::ssl::context& contex
         });
 
     return future;
-}
\ No newline at end of file
+}
+
+void SocketWrapper::close() {
```
Comment (shevmee) [inline include/SocketWrapper/src/SocketWrapper.cpp]: Name according to the code style

File: include/SocketWrapper/src/SocketWrapper.cpp [hunk 4]
```diff
@@ -106,4 +106,56 @@ std::future<void> SocketWrapper::startTlsAsync(boost::asio::ssl::context& contex
...
+    if (std::holds_alternative<std::shared_ptr<TcpSocket>>(socket_)) {
+        closeTcp();
+    } else if (std::holds_alternative<std::shared_ptr<SslSocket>>(socket_)) {
+        closeSsl();
+    }
+}
+
+bool SocketWrapper::is_open() const {
```
Comment (shevmee) [inline include/SocketWrapper/src/SocketWrapper.cpp]: Name according to the code style

File: include/SocketWrapper/src/SocketWrapper.cpp [hunk 5]
```diff
@@ -106,4 +106,56 @@ std::future<void> SocketWrapper::startTlsAsync(boost::asio::ssl::context& contex
...
+    if (is_tls_) {
+        auto ssl_socket = std::get<std::shared_ptr<SslSocket>>(socket_);
+        return ssl_socket && ssl_socket->lowest_layer().is_open();
+    } else {
+        auto tcp_socket = std::get<std::shared_ptr<TcpSocket>>(socket_);
+        return tcp_socket && tcp_socket->is_open();
+    }
+}
+
+void SocketWrapper::closeTcp() {
```
Comment (shevmee) [inline include/SocketWrapper/src/SocketWrapper.cpp]: Name according to the code style

File: include/SocketWrapper/src/SocketWrapper.cpp [hunk 6]
```diff
@@ -106,4 +106,56 @@ std::future<void> SocketWrapper::startTlsAsync(boost::asio::ssl::context& contex
...
+        }
+
+        tcp_socket->close(ec);
+        if (ec && ec != boost::asio::error::not_connected) {
+            std::cerr << "Error closing TCP socket: " << ec.message() << std::endl;
+        }
+    }
+}
+
+void SocketWrapper::closeSsl() {
```
Comment (shevmee) [inline include/SocketWrapper/src/SocketWrapper.cpp]: Name according to the code style

File: include/SocketWrapper/include/SocketWrapper.h [hunk 7]
```diff
@@ -32,9 +32,17 @@ class SocketWrapper {
...
+
+    void close();
+
+	bool is_open() const;
+	
 private:
 	std::variant<std::shared_ptr<TcpSocket>, std::shared_ptr<SslSocket>> socket_;
 	bool is_tls_;
+
+	void closeTcp();
```
Comment (shevmee) [inline include/SocketWrapper/include/SocketWrapper.h]: Name according to the code style

File: include/Server/src/Server.cpp [hunk 8]
```diff
@@ -56,13 +56,35 @@ void SmtpServer::Accept() {
     });
 }
 
+void SmtpServer::resetTimeoutTimer(SocketWrapper& socket_wrapper) {
```
Comment (shevmee) [inline include/Server/src/Server.cpp:60]: Name according to the code style

**Review by shevmee — COMMENTED** [review]
The code is well-structured and readable. Good job on maintaining consistency and following the coding standards. But take care about the code style.

File: include/CommandHandler/include/CommandHandler.h [hunk 9]
```diff
@@ -19,45 +19,44 @@ namespace ISXSC {
...
     boost::asio::ssl::context& ssl_context_;
-    std::unique_ptr<ISXMailDB::PgMailDB> data_base_;
-    MailMessageBuilder mail_builder_;
-    bool in_data_ = false;
+    std::unique_ptr<ISXMailDB::PgMailDB> m_data_base_;
+    MailMessageBuilder m_mail_builder_;
+    bool m_in_data_ = false;
     std::string connection_string_ = "postgresql://postgres.qotrdwfvknwbfrompcji:"
                                      "yUf73LWenSqd9Lt4@aws-0-eu-central-1.pooler."
                                      "supabase.com:6543/postgres?sslmode=require";
```
Comment (vsh51) [inline include/CommandHandler/include/CommandHandler.h]: Class member variable. Think of using m_ prefix

File: include/Server/src/Server.cpp [hunk 10]
```diff
@@ -28,41 +28,63 @@ SmtpServer::SmtpServer(boost::asio::io_context& io_context, boost::asio::ssl::co
...
+            socket_wrapper.Close(); 
+        } catch (const std::exception& e) {
+            std::cerr << "Exception in timeout handler: " << e.what() << std::endl;
+            ErrorHandler::handleException("Timeout handler", e);
+        }
+    });
+}
+
 void SmtpServer::handleClient(SocketWrapper socket_wrapper) {
     try {
```
Comment (vsh51) [inline include/Server/src/Server.cpp:80]: Naming issue

File: include/Server/src/Server.cpp [hunk 11]
```diff
@@ -117,6 +145,7 @@ void SmtpServer::saveData(const std::string& line, MailMessageBuilder& mail_buil
         }
     }
 }
+
 // Save the email message to a file
 void SmtpServer::tempSaveMail(const MailMessage& message) {
```
Comment (vsh51) [inline include/Server/src/Server.cpp:150]: Naming issue

**Review by vsh51 — COMMENTED** [review]
The code was well-structured, making the implementation of the timeout feature efficient, despite minor naming issues. Also I'd recommend paying extra attention to error handling in callbacks, it might be tricky and appear on testing stage. Generally code looks well

File: include/SocketWrapper/src/SocketWrapper.cpp [hunk 12]
```diff
@@ -108,26 +108,52 @@ std::future<void> SocketWrapper::startTlsAsync(boost::asio::ssl::context& contex
...
+		m_timeout_timer_->cancel(ec);
+		if (ec) {
+			std::cerr << "Failed to cancel timeout timer: " << ec.message() << std::endl;
+		}
+	}
+}
+
 void SocketWrapper::Close() {
-    if (std::holds_alternative<std::shared_ptr<TcpSocket>>(socket_)) {
+    if (std::holds_alternative<std::shared_ptr<TcpSocket>>(m_socket_)) {
```
Comment (shevmee) [inline include/SocketWrapper/src/SocketWrapper.cpp]: keep only m_socket(without underdash at the end)

File: include/SocketWrapper/src/SocketWrapper.cpp [hunk 13]
```diff
@@ -108,26 +108,52 @@ std::future<void> SocketWrapper::startTlsAsync(boost::asio::ssl::context& contex
...
+			if (!error) {
+				std::cout << "Client timed out." << std::endl;
+				this->Close();
+			}
+		});
+	}
+}
+
+void SocketWrapper::CancelTimeoutTimer() {
+	if (m_timeout_timer_) {
```
Comment (shevmee) [inline include/SocketWrapper/src/SocketWrapper.cpp]: underdash at the end of variable isn't implied by the code style

**Review by shevmee — COMMENTED** [review]
Good, but rename variables

**Pull request merged**